algebra/direct_limit.lean

```diff
@@ -0,0 +1,290 @@
+import algebra.ring order.lattice
+
+universes u v
+
+open lattice
+
+instance is_ring_hom.to_is_add_group_hom {α : Type*} {β : Type*} [ring α] [ring β]
+  (f : α → β) [is_ring_hom f] : is_add_group_hom f :=
+⟨λ _ _, is_ring_hom.map_add f⟩
+
+class directed_order (α : Type u) extends has_sup α, partial_order α :=
```

Comment:
This definition uses inaccurate terminology. I don't know how much of an issue this is. `has_sup` is clearly intended for situations where two objects have a supremum, or *least* upper bound. For directed sets (perhaps "directed types" would be better in this context, I am less sure about "directed orders") we definitely don't want to demand that two objects have a supremum, just a common upper bound. In particular the use of `sup` and the corresponding symbol is arguably misleading in this context.

Reply:
I agree - I don't think a directed order should be associated with any additional operations at all, just an existence claim for each pair of points in the type. I think in many situations there is no canonical choice for the witness, so we shouldn't bake in a particular one or else it will mess up equality of structures.

Reply:
So you would prefer an exists statement? But you've seen the proofs -- `calc   f (p ⊔ q) (p ⊔ q ⊔ (r ⊔ s) ⊔ (t ⊔ u)) ...`. Presumably there's some sensible way to make it so that the proofs can stay looking like this -- some noncomputable function with notation attached I guess. What would be good notation for this?

Reply:
I don't think that expression is needed in the first place. If you want larger suprema, you can prove that any list / finset of elements has an upper bound. Even in traditional mathematics it seems a bit weird to have a function here, unless you need it for a particular choicey construction.

Reply:
To put this into some context, I would like to use this direct limit stuff in the following situation: X is a top space, B is a basis of open sets, x is in X, and the system is the set of basic open sets containing x. The intersection of two basis elements need not be a basis element, but will contain such. I would define i <= j iff j is contained in i. So what Mario says (the bare existence statement) is exactly what comes out of the definition of a basis.

Reply:
@digama0 I don't understand. How do I make the function computable if I use existentials?